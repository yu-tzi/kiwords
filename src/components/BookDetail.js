import React from "react";
import { db } from "../utility/firebaseConfig"
import '../style/BookDetail.scss';
import Loading from './Loading.js';


class BookDetail extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      allCards:[],
      cards: "",
      wordDetail: false,
      nowWord: "",
      wordRenew: false,
      page: 0,
      lastValue: 0
    };
    this.renderWords = this.renderWords.bind(this)
    this.renderCards = this.renderCards.bind(this)
    this.popWord = this.popWord.bind(this)
    this.renderPage = this.renderPage.bind(this)
    this.renderPageDetail = this.renderPageDetail.bind(this)
    this.changePage = this.changePage.bind(this)
    this.convertWordString = this.convertWordString.bind(this)
    this.renewWordList = this.renewWordList.bind(this)
  }

  componentDidMount() {
    let url = /* "https://kiwords-c058b.web.app/details/159592106485385aFFbQvKxZmidyhpfaKGrl2uPL2?SAT%202500" */window.location.href

    let bookID
    let bookName
    for (let i = 0; i < url.split("/").length; i++) {
      let target = url.split("/")[i]
      bookID = target.split("?")[0]
      bookName = decodeURI(target.split("?")[1])
    }

    let cards = []
    let verifyCount = 0

    if (bookID.length > 0) {
      db.collection("books")
        .doc(bookID)
        .get()
        .then((doc) => {
          if (doc.exists) {
            this.setState({ allCards: doc.data().cards })
            this.setState({ bookOwner: doc.data().created })
          } else {
            console.log("No such document!");
          }
        })
        .then(() => {
          if (this.state.allCards.length > 0) {
            //get page number
            if (this.state.allCards.length % 8 === 0) {
              this.setState({ page: Math.floor(this.state.allCards.length / 8) })
            } else {
              this.setState({ page: Math.floor(this.state.allCards.length / 8) + 1 })
            }
            //get first page words
            let lastNum
            if (8 > this.state.allCards.length) {
              lastNum = this.state.allCards.length % 8
            } else {
              lastNum = 8
            }
            for (let i = 0; i < lastNum; i++) {
              cards.push(this.renderCards(this.state.allCards[i], i))
              verifyCount++
            }
            if (verifyCount = this.state.allCards.length) {
              this.setState({ cards: cards })
            }
          } else {
            this.setState({ cards: [] })
            console.log("No cards :(")
          }
        })
        .catch((error) => {
          console.log("Error getting document:", error);
        })
    }
  }

  componentDidUpdate() {

    let verifyCount = 0
    let isContentChange = this.state.cards.length >= 1 && this.state.wordRenew
    if (isContentChange) {
      let cards = []
      let lastNum
      let isLastPage = this.state.lastValue + 8 > this.state.allCards.length
      let isNoWords = this.state.lastValue < 1 && this.state.allCards.length < 1
    

      if (this.state.allCards.length % 8 === 0) {
        this.setState({ page: Math.floor(this.state.allCards.length / 8) })
      } else {
        this.setState({ page: Math.floor(this.state.allCards.length / 8) + 1 })
      }

      if (isLastPage && this.state.allCards.length > 0) {
        lastNum = this.state.lastValue + this.state.allCards.length % 8
      } else if (isLastPage && isNoWords) {
        cards = []
        this.setState({ cards: cards })
        this.setState({ wordRenew: false })
        } else {
        lastNum = this.state.lastValue + 8
      }
      
      for (let i = this.state.lastValue; i < lastNum; i++) {
        cards.push(this.renderCards(this.state.allCards[i], i))
        verifyCount++
      }
      if (verifyCount = this.state.allCards.length) {
        this.setState({ cards: cards })
        this.setState({ bookOwner: this.state.bookOwner })
        this.setState({ wordRenew: false })
      }
    }
  }

  renderCards(data, i) {

    let synonyms = this.convertWordString(data.synonyms)
    let antonym = this.convertWordString(data.antonym)
    let isNowWord = parseInt(this.state.nowWord) === i

    return (

      <div key={i}>
        
        <div id="wordBlock"
          onClick={(e) => {
            this.popWord(e)
          }} className={i}>
          <div className={i}>{data.word}
          </div>
          <img src="images/trash.png" id="trashCan" className={i}
            onClick={(e) => {
              this.deleteWord(e)
            }}>
          </img>
        </div>

        <div id="popWordDetail" className={i}
          style={{
            display: this.state.wordDetail && isNowWord ? "block" : "none"
          }}>
          <div className="popWordDetailTitle">Meaning
          </div>
          <div className="popWordDetailContent">
            {data.meaning}
          </div>
          <div className="popWordDetailTitle">Synonyms
          </div>
          <div className="popWordDetailContent">
            {data.synonyms.length > 0 ? synonyms : "No synonyms"}
          </div>
          <div className="popWordDetailTitle">Antonym
          </div>
          <div className="popWordDetailContent">
            {data.antonym.length > 0 ? antonym : "No antonym"}</div>
        </div>

      </div>
    )
  }

  convertWordString = (str) => {
    let result = ""
    if (typeof (str) === "object") {
      for (let i = 0; i < str.length; i++) {
        if (i === 0) {
          result += str[i]
        } else {
          result += " ," + str[i]
        }
      }
    } else {
      result = str
    }
    return result
  }

  renderWords() {
    return (
      <div className="wordCardBox">
        {this.state.cards}
      </div>
    )
  }

  popWord(e) {
    e.stopPropagation()
    this.setState({ wordDetail: !this.state.wordDetail })
    this.setState({ nowWord: e.target.className })
    this.setState({wordRenew: true})
  }

  deleteWord(e) {

    let all = []
    let className = e.target.className
    e.stopPropagation()

    for (let i = 0; i < this.state.allCards.length; i++) {
      if (parseInt(className) === i) {
      } else {
        all.push(this.state.allCards[i])
      }
    }
    this.renewWordList(all)
  }

  renewWordList(data) {

    let url = /* "https://kiwords-c058b.web.app/details/159592106485385aFFbQvKxZmidyhpfaKGrl2uPL2?SAT%202500" */ window.location.href
    let bookID

    for (let i = 0; i < url.split("/").length; i++) {
      let target = url.split("/")[i]
      bookID = target.split("?")[0]
    }
    db.collection("books").doc(bookID)
      .update({
      cards: data
      })
      .then(() => {
        if (data.length <= this.state.lastValue && this.state.lastValue > 0) {
          this.setState({ lastValue: this.state.lastValue - 8 })
        } 
      })
      .then(() => {
        this.setState({ allCards: data })
        this.setState({ wordRenew: true })
      })
      .catch((err) => {
        console.log(err)
      })
  }


  renderPage() {
    let All = []
    for (let i = 0; i < this.state.page; i++) {
      All.push(this.renderPageDetail(i))
    }
    return (
      <div className="pageLeaf">
        {All}
      </div>
    )
  }

  renderPageDetail(i) {
    return (
      <div className="page" key={i} onClick={() => { this.changePage(i) }}>
        {i + 1}
      </div>
    )
  }

  changePage(i) {
    let num = parseInt(i) * 8
    this.setState({ lastValue: num })
    this.setState({ wordRenew: true })
  }

  



  render() {

    let rootURL = window.location.href.substr(0, window.location.href.indexOf("/", 9))
    let url = /* "https://kiwords-c058b.web.app/details/159592106485385aFFbQvKxZmidyhpfaKGrl2uPL2?SAT%202500"  */window.location.href
    let bookID
    let bookName

    for (let i = 0; i < url.split("/").length; i++) {
      let target = url.split("/")[i]
      bookID = target.split("?")[0]
      bookName = decodeURI(target.split("?")[1])
    }

    return (
      <div className="bookDetailBody">
        
        <div className="upperPart">
          <div className="bookDetailTitle">
            {bookName}
          </div>
          <div className="upperbtn">
            <div className="viewAllBookBtn"
              onClick={() => {
                window.location.href = (rootURL + '/wordbooks')
              }}>View other wordbooks
            </div>
            <div className="addWordsBtn"
              onClick={(event) => {
                event.stopPropagation(),
                window.location.href = ('https://kiwords-c058b.web.app/addwords?' + bookID + "&" + bookName)
              }}>Add more words
            </div>
          </div>
        </div>

        <div className="wordSub"
          style={{
            display: this.state.cards.length > 3 ? "block" : "none"
          }}>Words in this book
        </div>

        {
          this.state.cards.length > 0 ?
          <div className="wordCardBox">
              {this.state.cards}
          </div>
           : 
          <div>
              {
                typeof (this.state.cards) === "object" ? 
                  <div className="noWordsBlock">
                    <div className="noWordsTitle">There's no words in this wordbook
                    </div>
                    <div className="noWordSubTitle">Try adding some words ! Add words with more then 2 synonyms and antonyms to unlock quiz function.
                    </div>
                    <div className="noWordBtn"
                      onClick={() => {
                        window.location.href = ('https://kiwords-c058b.web.app/addwords?' + bookID + "&" + bookName)
                      }}>Add words
                    </div>
                  </div>
                  :
                  <div className="noWordsBlock">
                    <Loading loadingMini={false} />
                  </div>
              }
          </div>
        }

        {this.renderPage()}

        <div className="bottomPart"
          style={{
            display: this.state.cards.length > 1 ? "flex" : "none"
          }}>
          <div className="takeQuizTitle">Familiar with all these words ?
          </div>
          <div className="takeQuizBtn"
            onClick={(event) => {
              event.stopPropagation(),
              window.location.href = ('https://kiwords-c058b.web.app/quiz?' + bookID + "&" + bookName)
            }}>Take a quiz
          </div>
        </div>

        <div className="bottomBlank">
        </div>

      </div>
    )
  }
}




export default BookDetail