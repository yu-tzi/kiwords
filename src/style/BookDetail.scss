.bookDetailBody{
  color: white;
  width: 80%;
  margin: auto;
  max-width: 1000px;
  
}

#wordBlock{
  display:flex;
  justify-content: space-between;
}

.bookDetailTitle{
  text-align: center;
  font-size: 28px;
  line-height: 38px;
  margin-bottom: 0px;
}

.wordSub{
  color: #3ccfcf;
  font-weight: 600;
}

.upperPart{
  display: flex;
  justify-content: space-between;
  padding: 30px 0 15px 0;
  width: 80%;
  margin: auto;
}

.takeQuizBtn{
    font-size: 16px;
    background-color: orange;
    border-radius: 3px;
    padding: 8px 10px;
    color: #303545;
    text-align: center;
    line-height: 22px;
    margin-bottom: 0px;
    width: 100px;
    margin-top: 15px;
    transition: 0.5s;
    cursor: pointer;
      }

.noWordsBlock{
  width: 80%;
    margin: auto;
    max-width: 1000px;
    margin-top: 30px;
    margin-bottom: 50px;
    /* justify-content: center; */
    /* align-items: center; */
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-top: 100px;
}

.noWordsTitle{
  color: white;
    font-size: 30px;
}

.noWordBtn{
  color: #303545;
    margin-top: 30px;
    background-color: #3ccfcf;
    height: 40px;
    width: 200px;
    border-radius: 5px;
    text-align: center;
    line-height: 40px;
    transition:0.5s;
}

.noWordSubTitle{
      color: white;
    font-weight: 200;
    padding-top: 15px;
}

.noWordBtn:hover{
  color: #3ccfcf;
  background-color: #303545;
}

.bottomBlank{
  height:80px;
  width:80px;
}

.addWordsBtn{
  font-size: 16px;
    background-color: orange;
    border-radius: 3px;
    padding: 8px 10px;
    color: #303545;
    text-align: center;
    line-height: 22px;
    margin-bottom: 0px;
    margin-left: 20px;
    transition: 0.5s;
}

.upperbtn{
  display: flex;
}

.viewAllBookBtn{
      font-size: 16px;
    background-color: orange;
    border-radius: 3px;
    padding: 8px 10px;
    color: #303545;
    text-align: center;
    line-height: 22px;
    margin-bottom: 0px;
    margin-left: 20px;
    transition: 0.5s;
}


.wordSub{
  color: #3ccfcf;
  font-weight: 600;
  padding-top: 20px;
  width: 80%;
  margin: auto;
}

.wordCardBox{
  margin: auto;
  width: 80%;
}

#wordBlock{
  border-radius: 3px;
  background-color: #24293c;
  padding: 7px;
  margin: 10px 0 5px;
  border-color: #24293c;
  border-width: 3px;
  border-style: solid;
}

#popWordDetail{
  word-wrap:break-word;
  background-color: #181d30;
  display: block;
  padding: 1px 8px 10px;
}

.pageLeaf{
  display: flex;
  align-items: center;
  margin-top: 50px;
  width: 100%;
  justify-content: center;
}

.page{
  border-radius: 5px;
  background-color: #495278;
  padding: 10px;
  font-size: 20px;
  text-align: center;
  margin: 0px 10px;
  transition: 0.5s;
}

.page:hover{
  background-color: #24293c;
}

.popWordDetailContent{
  font-weight: 100;
}

.popWordDetailTitle{
  margin-top: 10px;
  margin-bottom: 5px;
}

#wordBlock:hover{
  border-bottom-color:orange;
}

#trashCan{
  height: 15px;
}

.bottomPart{
  width: 80%;
  margin: auto;
  margin-top: 30px;
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #495278;
  height: 150px;
  align-content: center;
  justify-content: center;
  border-radius: 5px;

}

.takeQuizTitle{
  font-size: 22px;
  

}

.addWordsBtn:hover,.viewAllBookBtn:hover,.takeQuizBtn:hover{
  background-color: #303545;
  color: orange ;
}


@media (max-width: 450px) {

.takeQuizTitle{
  width: 80%;
}

.upperPart{
  flex-direction: column;
}

.upperbtn{
  display: block;
}

.viewAllBookBtn,.addWordsBtn{
  margin: 10px auto;
  width: 70%;
}

}