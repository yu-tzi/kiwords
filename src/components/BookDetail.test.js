import BookDetail from "./BookDetail.js";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import React from "react";
Enzyme.configure({ adapter: new Adapter() });

it("snapshot tests", () => {
  const tree = renderer.create(<BookDetail />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("BookDetail render exist", () => {
  const wrapper = shallow(<BookDetail />);
  expect(wrapper.exists()).toBe(true);
});

it("count lastValue function", () => {
  const component = mount(<BookDetail />);
  component.instance().changePage(1);
  expect(component.state("lastValue")).toBe(8);
  component.instance().changePage(3);
  expect(component.state("lastValue")).toBe(24);
  component.instance().changePage(0);
  expect(component.state("lastValue")).toBe(0);
});

it("pageLeaf length based on page state", () => {
  const component = mount(<BookDetail />);
  component.setState({ page: 3 });
  expect(component.find(".page").length).toBe(3);
  component.setState({ page: 0 });
  expect(component.find(".page").length).toBe(0);
});
