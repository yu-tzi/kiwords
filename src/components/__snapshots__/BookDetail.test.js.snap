// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests 1`] = `
<div
  className="bookDetailBody"
>
  <div
    className="upperPart"
  >
    <div
      className="bookDetailTitle"
    >
      undefined
    </div>
    <div
      className="upperbtn"
    >
      <div
        className="viewAllBookBtn"
        onClick={[Function]}
      >
        View other wordbooks
      </div>
      <div
        className="addWordsBtn"
        onClick={[Function]}
      >
        Add more words
      </div>
    </div>
  </div>
  <div
    className="wordSub"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Words in this book
  </div>
  <div
    className="noWordsBlock"
  >
    <div
      className="noWordsTitle"
    >
      There's no words in this wordbook
    </div>
    <div
      className="noWordSubTitle"
    >
      Try adding some words ! Add words with more then 2 synonyms and antonyms to unlock quiz function.
    </div>
    <div
      className="noWordBtn"
      onClick={[Function]}
    >
      Add words
    </div>
  </div>
  <div
    className="pageLeaf"
  />
  <div
    className="bottomPart"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="takeQuizTitle"
    >
      Familiar with all these words ?
    </div>
    <div
      className="takeQuizBtn"
      onClick={[Function]}
    >
      Take a quiz
    </div>
  </div>
  <div
    className="bottomBlank"
  />
</div>
`;
